lib/matplotlib/dates.py

```diff
@@ -1393,8 +1393,10 @@ def __init__(self, base=1, month=1, day=1, tz=None):
                          'hour':   0,
                          'minute': 0,
                          'second': 0,
-                         'tzinfo': tz
                          }
+        # Note that tzinfo does not work with pytz timezones, and
```

Comment:
@pganssle : the way that tz info was put in before was to do `vmin.replace(year=ymin, **self.replaced)`, but `pytz` doesn't work properly (at all) with `replace`.  So we need a test with pytz to test that this code works.  

To be honest, we roundtrip most dates (datetime-> ordinal -> datetime) so the fact that pytz doesn't do the right thing w/ `tzinfo` isn't that big a deal practically, but it would be nice if it worked.

I guess we could add another test that doesn't use pytz, but I think pytz should be tested on this code path.

Reply:
`replace(tzinfo=tz)` and `astimezone(tz)` do two very different things.

If you think of time zones as units for datetimes, `replace(tzinfo=tz)` is like getting a value that says "1 ft", scratching off the "ft" and adding "m" to get "1m". `astimezone(tz)` is like getting something that says 1ft and converting it to meters to get `0.3m`.

Generally speaking, `replace` is used if you know that the local portion of the datetime is correct and want to say what time zone it's in. `astimezone` is used if you know that the absolute time represented by the datetime is correct and you want to convert it to a different local time.

In Python >=3.6, `astimezone` can be used on naive datetimes, but what that does is that it interprets the naive datetime as *local time*, and converts that to the time zone of interest. What that means is that if you are in Tokyo and give this a naive datetime like `datetime(2018, 1, 1)` and have `YearLocator(tz=tz.gettz("America/New_York")`, that's going to get converted to `2017-12-31 10:00:00-05:00`, which is probably not what you wanted.

Reply:
I'll note that I'm [well aware of the limitations of `pytz`](https://blog.ganssle.io/articles/2018/03/pytz-fastest-footgun.html).

I think probably it's best to use two tests, one with `pytz` (skipped if `pytz` is not present) and one with `dateutil`. If you intend for the interface to look the same, you can likely parametrize the test function with a conditional skip on `pytz`.